### PR TITLE
[WIP] API Docs

### PR DESCRIPTION
**Pull request by BrunnerLivio**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/docs.nestjs.com/blob/master/CONTRIBUTING.md

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] Other... Please describe:
```

## What is the current behavior?
No detailed API documentation

Issue Number: #249 

## What is the new behavior?

Adds an API page which displays detailed information of all the exposed classes / constants / interfaces / etc. of NestJS. 

simplescreenrecorder20190513

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information

- The code is highly inspired by the angular repository
- To generate the doc pages from the source code it uses a tool called  [dgeni](https://github.com/angular/dgeni).
- **This PR will introduce git submodules, so dgeni can generate the pages from the code**
- It is far from finished - Just want to involve you into the process asap :)
- This PR will get big without a doubt :/
- URLs in code (cross-references of types for example) will get overwritten by Prism highlighter - we need to find a workaround / other library, because URL in code would be super helpful for the user.

### CODE REVIEW

**Comment by BrunnerLivio** [comment]
nestjs-api-docs

It is getting better and better ;)

**Comment by 1valdis** [comment]
Any ETA on that? I'm coming from Sequelize which has full API reference, so it's hard for me to understand what I'm working with without API reference :(

**Comment by priyashpatil** [comment]
@kamilmysliwiec what's eta for this. Waiting for it.

**Comment by BrunnerLivio** [comment]
Waiting for #969, so I can implement dynamic routing and then revisit this PR. It depends on all those things.

**Comment by tonivj5** [comment]
@BrunnerLivio https://github.com/nestjs/docs.nestjs.com/pull/969 was reverted. Is there any plan to continue with this PR? Really good work! :rocket:

**Comment by knownasilya** [comment]
Really interested in this as well.

**Comment by BrunnerLivio** [comment]
I'd love to revisit this PR, though my day-to-day job is taking all of my energy at the moment.

Though in case you're interested to contribute to it, here are some hints about my strategy;

The main reason why this PR is frozen is that it needs quite some underlying change related to **routing**. My aim was to basically implement a similar architecture as the Angular documentation. 

In summary, the way the Angular documentation works is that it generates documents from various data source (e,g, markdown or typescript files) and translates it to HTML. This is mostly already done in this PR -- meaning you can expect that there are static HTML files that contain all the information about all the classes which exist in NestJS.

Now the part which I was unable to finish so far:
This content needs to be dynamically displayed on the documentation. Meaning if the user for example requests `docs.nestjs.com/api/core/logger`, the documentation will need to look after a local file in `static/core/logger.html` and place the contents inside the main body of the website.

Sounds easier than it sounds. On of the reason for that is that the HTML files can contain Angular components (e.g. for displaying the code). Angular components can not load dynamically when you load an external HTML content and place it via e.g. `innerHTML` into the DOM. What the Angular team has done, they converted all their Angular components which can be placed into the DOM to Web Components using Angular Elements. That is why I started for example #1214. 

On top of that, the Angular team has written its own router, since `@angular/router` does not provide the capability of having such dynamic content.  They even dynamically load the Web Components, depending on which HTML document will be displayed on the docs. Meaning the user does not need to load unnecessary JS. But that is not mandatory to implement in the case of the NestJS docs in my opinion.

Now steps forwards;

I think it would make sense to have a closer look at [Scully](https://scully.io/). I haven't gotten my hands dirty with it yet, but in case it works similar to GatsbyJS, then it is definitely something we should use in order to make this PR happen. With that we could by-pass all of the routing stuff mentioned above.

So in short; I'd love to see a PoC with Scully and see whether this person could dynamically create pages using HTML content provided by this PR with Angular components from the NestJS docs.

**Comment by BrunnerLivio** [comment]
Someone just asked me privately how to run this branch locally in 2022.

Here are updated instructions:

1. Some dependencies are quite outdated and use http-based connection with Github. To fix this, add this to your .gitconfig. [Read more here](https://stackoverflow.com/questions/45433130/npm-install-gets-stuck-at-fetchmetadata)

```
[url "https://"]
    insteadOf = git://
```

2. Execute these commands in your Terminal (make sure you have nvm installed):

```
git clone -b feature/api-docs git@github.com:BrunnerLivio/docs.nestjs.com.git
nvm use 10
cd docs.nestjs.com
npm ci
cd sources/
git clone git@github.com:nestjs/terminus.git 
# Clone other nestjs/* repos in case you need them
cd ..
npm start
```

**Comment by vahidvdn** [comment]
Any update on this?

**Comment by joaquinnunez** [comment]
i guess, i'll have to clone the project and `git grep` my way to read the api reference, sad this is not live

**Comment by knownasilya** [comment]
Is nestjs abandoned?

**Comment by jmcdo29** [comment]
> Is nestjs abandoned?

@knownasilya what gave you that idea? Kamil is literally working on the v10 branch and has been for the past few weeks. Just because this API docs that was an experimental branch hasn't been merged does not mean it's anywhere **close** to abandoned.

**Comment by knownasilya** [comment]
Wasn't aware of v10 work, seems that api reference docs would be ultra high priority for a project like this.

**Comment by knownasilya** [comment]
BTW @jmcdo29 you have always been very helpful whenever I had questions, so thanks!

**Comment by 1valdis** [comment]
Question is not just having the list of public API; without proper description for each entry it's just a small part of the solution. All of the APIshould be properly documented as well. For example, there's a very interesting thing in Core called DiscoveryService. How a developer is supposed to learn how it works and what it can be used for? By reading the source? Or other repositories which are built upon it? That's suboptimal.

**Comment by 1valdis** [comment]
Regarding you comment above, Kamil: I personally didn't know one could press `.` and open and editor; and also I didn't know I should search for `@publicApi` and that would be the list of everything that's intended for users. Is this documented somewhere other than in this issue at this point?

**Comment by knownasilya** [comment]
Thanks for clarifying @Tony133, looking forward to it

**Comment by Tony133** [comment]
Hi all 😀 👋🏻, as promised, here is the link to access the Api Docs v2.0-beta for NestJS:

Link: https://api-references-nestjs.netlify.app 🔥 🚀 

Lately we have also added a drop-down menu to switch between NestJS version 9 and version 10. 🙌

I add below some technical notes as described in the previous post:

- The following official packages:

```
    * @nestjs/graphql
    * @nestjs/swagger
    * @nestjs/config
    * @nestjs/passport
    * @nestjs/jwt
    * @nestjs/mapped-types
    * @nestjs/cqrs
    * @nestjs/typeorm
    * @nestjs/sequelize
    * @nestjs/mongoose
    * @nestjs/serve-static
    * @nestjs/schedule
    * @nestjs/bull
    * @nestjs/elasticsearch
    * @nestjs/event-emitter
    * @nestjs/axios
```

 do not have the `@publicApi` annotation, I added them manually.

Currently it is necessary to maintain and synchronize the forks of these packages (which I am managing), because we need to add the JSDoc @publicApi comments, so that the API generator knows what is considered an export that is part of the API, and it is a bit problematic to maintain.

Currently only the following packages have the JSDoc `@publicApi` annotation:

```
 - @nestjs/terminus
 - @nestjs/throttler
 - @nestjs/cache-manager 
 ```

In addition to the nest repository (https://github.com/nestjs/nest), within the monorepository are the general and essential NestJS packages, which are as follows:

```
- @nestjs/common
- @nestjs/core
- @nestjs/microservices 
- @nestjs/platform-express
- @nestjs/platform-fastify
- @nestjs/websockets
- @nestjs/platform-ws
- @nestjs/platform-socket.io
- @nestjs/testing
```

I conclude by leaving you with this question:

Which official packages do you find most useful to have in an Api Docs for NestJS ?

- All packages  ?

- Just a few (e.g., excluding those dedicated to ORMs, just as a quick example) ?


This point is an open question on which I think it is fair for the entire community of developers using NestJS to give their input, as well as members of the core team, to decide which way to go. 

Most likely this will be the last chance for NestJS to have its own Api Reference, not all frameworks have it it is true, but I think it is very useful for those who use NestJS and also for "newcomers" who are using it. 

There is always a saying, "Some 'opportunities' when they pass you have to seize them on the fly." 
NestJS is a great opportunity, let's not waste it.  💪

This "poll and discussion" however will go on for a while until the next major release of NestJS (v11.x), then we will see if we extend it further. 

Mark this date: July 7, 2023, NestJS to a working prototype of Api Docs as a great starting point, hoping that one day it will become the official one. 😻 ♥️

I have gone on and on in this post, but I like to specify things well 😄. I hope you can appreciate this work started by @BrunnerLivio 4 years ago and carried on by me always with his supervision 🙌, and that it can be useful during your projects with NestJS.😻 🙏🏻

P.s: We are considering whether to open a post on the Discord support channel to move all feedback maybe there, we are still considering with @BrunnerLivio how to proceed. In case of updates we will write you! 😉👍

**Comment by Tony133** [comment]
Two small updates 👀:

In these weeks after the launch, I also made a DEV.to post about this prototype see here: https://dev.to/tony133/api-docs-v20-beta-prototype-for-nestjs-3jdj , so it would be even more visible 🚀 🌎, and I also created a thread in the discord support channel  (direct link to the thread): https://discord.com/channels/520622812742811698/1128439541326417962 💪

That is all 😉 👍

**Comment by c3cris** [comment]
Every framework I know documents its API, this should be a priority.

**Comment by Tony133** [comment]
As of today, find Api Docs for NestJS in the official documentation ( see screenshot ) 🔥 🚀

api-reference
